### ISSUE TITLE
Android 12 _sans Chinese text display garbled

### ISSUE THREAD

**Issue by a9848840** [airsdk/Adobe-Runtime-Support#1299]
Hi 

I use text field default system word _sans
Chinese text display garbled    

"_sans"
"_serif"
"_typewriter"
also garbled □□□...

I googled Android 12 sans, this part has some adjustments, 
Can you help adjust the Chinese display as soon as possible? 

thank you

Android 12
AirSDK : 33.1.1.633 / 33.1.1.554

**Comment by Ender22** [comment]
This is a huge issue for my app as well, happy you guys are looking into it already, thanks~

**Comment by funyamora** [comment]
I have the same problem in Japanese including Chinese Characters.
But using Flash Text Engine without TextField, characters don't garbled.
https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/text/engine/package-detail.html

**Comment by Devin-Licastro** [comment]
Glad to hear someone is looking into this. Please ensure it is a priority. We have a lot of users who need our app's interface in Chinese and it's been over a month where some of them can't use our app. 

Also, this should be labeled as a bug, not a feature right? All the Chinese text in air apps displaying as [] when users update android is certainly a bug, isn't it?  Just hoping it doesn't get a lower priority when labeled this way.

**Comment by ajwfrost** [comment]
@Devin-Licastro are you seeing this issue on anything other than devices running Android 12? Would definitely class it as a bug if it's happening on Android 11 .. 

Anyway: we just spoke on this and I'm hoping we can get a beta release uploaded within the next 7-10 days with all the Android 12 / API 31 changes.

**Comment by Devin-Licastro** [comment]
No only on Android 12. 
Thanks for the details about the coming release.

**Comment by wmtj** [comment]
I hope Harman will fix this bug as soon as possible. @ajwfrost

**Comment by a9848840** [comment]
@ajwfrost 

Please help
grateful

**Comment by ajwfrost** [comment]
Yes we're working on the updates needed here: it's actually causing a lot more work than we thought, because we're having to update the gradle version that's used for building all of the AIR runtimes and supporting tools for the Android targets, and the original build files hadn't been updated for quite some time.. but we're making progress and hoping to get this out within the next week or two.

thanks

**Comment by a9848840** [comment]
@ajwfrost

Is it possible before Christmas?

thanks

**Comment by ajwfrost** [comment]
@a9848840 very possibly - we have it building now so need to do some other related updates and then kick off the release process..

**Comment by a9848840** [comment]
@ajwfrost

I am looking forward to this version, and I hope you can tell me if there is an update, 

thank you

**Comment by ajwfrost** [comment]
Yes so we'd got it all building and sorted out the last issues, with Android 12 / SDK 31; our earlier checks using this version suggested that it would just fix this font issue .. but having created the build, it appears this is not the case....

So we're checking it again, I hope to get an update tomorrow. We'll be going ahead with a release (at least making it available, beta-style) so that folk can get the update for this and the other changes as needed, but I'm not sure whether it will have solved this one I'm afraid. If it is just a configuration issue, which I'm hoping, then it may be that we can have this sorted out with the new build, but I don't yet have a final answer..

thanks

**Comment by ajwfrost** [comment]
Update on this issue .. we had originally believed that by updating our build system - and the resource files used in the AIR SDK - to those from Android 12 / SDK 31, that the fonts would then work properly. But this doesn't seem to be the case..

What it currently looks like is that the Android 12 requirement to use the "Roboto" font as the default sans-serif font is resulting in a lack of support for Chinese characters where there is no automatic fallback. We may need to parse the provided strings to check for characters and to then request a different typeface from the system where the requested glyphs are available...

**Comment by ajwfrost** [comment]
Hi again... can I just confirm here what you're actually trying? There are a number of variables involved and it seems to us that it may be working in some scenarios.

The AS3 code we're using as a text case is pretty simple:
```
			_tfmt = new TextFormat("_sans", 30, 0);
			_txt = new TextField();
			_txt.defaultTextFormat = _tfmt;
			_txt.text = "Testing 测试  測試 characters";
			addChild(_txt);
```
In other words, a normal TextField object, with a TextFormat that has "_sans" as the font name.

Then when building it, there are a number of values that we use in the gradle settings that depend on what you have installed... e.g.
```
    compileSdkVersion 30
    buildToolsVersion "30.0.3"
    defaultConfig.targetSdkVersion 30
```
You can change these by changing what Android build tools and platforms you have downloaded, so e.g. in your Android SDK manager if you've downloaded build-tools 31 then you'll get
`    buildToolsVersion "31.0.0"`
and if you've downloaded Android "S" then you'll get
`    compileSdkVersion 30`
You can also change the `targetSdkVersion` via your application descriptor file: it defaults to 30 currently, and with our ongoing updates it would change to 31, but you can do that manually anyway.

The problem I have is that when I try this with a recent AIR SDK, I'm seeing the Chinese text appear (Google Pixel 5, Android 12). Even with '30' set for everything..
Are you able to upload a small APK file that is exhibiting the problem and we'll try to analyse it to see what the problems are? And are you also just able to re-check on the latest SDK version, in case you're still using 33.1.1.633?

thanks

**Comment by a9848840** [comment]
Hi
@ajwfrost 
apk [(https://reurl.cc/Mknnrm)] using  33.1.1.554


I will also test AIR 33.1.1.713
Because the build process has changed
Need to adjust

thanks for the reply

**Comment by wmtj** [comment]
Now, more and more phones are upgrading to Android 12, This bug is becoming a disaster for my app. I ran a series of tests and invited my users to help me test it:

The result: Pixel 3/XL, Pixel 4/XL, MI 10S, MI 11/Ultra, OPPO Reno5 Pro+ 5G ... All, does not work!

AIR SDK :  33.1.1.713
Addroid API Level : 30
Addroid SDK : android-31

The code: 
`private function addTextFields():void 
		{
			var _tfmt:TextFormat = new TextFormat("_sans", 60, 0);
			_tfmt.align = TextFormatAlign.CENTER;
			
			var _txt1:TextField = new TextField();
			_txt1.defaultTextFormat = _tfmt;
			_txt1.text = "Testing Chinese characters.";
			_txt1.width = _txt1.textWidth + 4;
			_txt1.x = 100;
			_txt1.y = 400;
			
			var _txt2:TextField = new TextField();
			_txt2.defaultTextFormat = _tfmt;
			_txt2.text = "简体中文测试，繁體中文測試。";
			_txt2.width = _txt2.textWidth + 4;
			_txt2.x = 100;
			_txt2.y = 600;

			var _txt3:TextField = new TextField();
			_txt3.defaultTextFormat = _tfmt;
			_txt3.text = "is 'Roboto' Font exist ? -- " + isFontExist("Roboto", true);
			_txt3.width = _txt3.textWidth + 4;
			_txt3.x = 100;
			_txt3.y = 800;
			
			addChild(_txt1);
			addChild(_txt2);
			addChild(_txt3);
`

The screenshots & apk:

[Screenshots. ](https://www.ziwei001.com/update/beta/test/Screenshots.zip)
[The APK.](https://www.ziwei001.com/update/beta/test/Android12ChineseTest_API30.apk)

**Comment by wmtj** [comment]
This bug has not been solved for more than 2 months. Please pay more attention to it and do not abandon non-English users, please. ... @ajwfrost

**Comment by ajwfrost** [comment]
Hi

We've been working on this, there are a number of challenges here that are due to how the text drawing was originally implemented (and curiously, I think this may also be behind why there are also emoji limitations on Android..). We had thought at first that this was to do with a change in Android 12 itself and needed to update our build system and compilation/target API levels to Android SDK level 31 - the release for which is just uploaded as a beta - but this then turned out not to affect it.

We now need to update the mechanism needed for font drawing which is taken from Android 4.0 - and for compatibility we need to build Android 4 in order to do this (it uses C++ code that's not part of the NDK...), and this is now causing us some problems, so we're looking at workarounds to maybe split the supported rendering mechanism depending on what version of Android the device is running.

We're also investigating how bad it would be if we switched to actually using the Android SDK to do the text drawing. This would be the most compatible and straightforward solution, but there are no NDK APIs for these sorts of operations so there will be a hit to call back into Java (which will then call back into the underlying C++ implementation). The plus side is that it would behave a lot more like 'normal' Android applications would..

So yes, apologies that this has dragged on, we are working on it though!

One quick thought that's just occurred to me: since the problem you're seeing is due to the Roboto font not supporting these characters, and the normal fallback mechanism is the part that's actually failing within the AIR code, could you try specifying a font that already supports these characters such as "NotoSansCJK"? (i.e. rather than "_sans" which translates to "Roboto", can you set "NotoSansCJK" as the font family?) - I'll see if one of the guys here can check this out or try a few of the different fonts, the file for NotoSansCJK appears to not be a normal truetype font format so I don't actually know whether the code from Android 4 would cope with it..


thanks

**Comment by wmtj** [comment]
> 

Does the 'quick thought' means : new TextFormat("Noto Sans CJK SC", 60, 0) , like this?  I have tested it.

The phone for testing : 
Pixel 3 XL (Android 12) 
MI K40 (MIUI 12.5.19, Android 11)

No matter what font family I set, it looks the same. In other words, changing the font family name doesn't work at all and the display is always same with _sans.

[Screenshots2.zip](https://www.ziwei001.com/update/beta/test/Screenshots2.zip)
[src.zip](https://www.ziwei001.com/update/beta/test/src.zip)

**Comment by wmtj** [comment]
Another information : 

I bought cloud device testing service form testin.cn , there are two categories of Android 12 phones : "12" and "12(Android S)". All the phones under "12(Android S)" work fine, but all the phones under "12" does not work (garbled □□□).
I asked their customer service, they only told me there are minor version number differences of the android 12. Detailed version number information is not available now. I hope you found this information helpful.

[Screenshots3.zip](https://www.ziwei001.com/update/beta/test/Screenshots3.zip)

**Comment by mucoci** [comment]
I have the same problem, but I solved it temporarily...

My players report this bug to me, too
and their phones are all Pixel series + android 12
finally,I guess ....it must lack certain a font...

then, I embedded a Font in my game...and use it in TextField(use .embedFonts = true)
all the problem is solved...

if you guys  are in a hurry, you can try this method first

**Comment by Devin-Licastro** [comment]
Just want to add that for my project, we have too many languages with glyphs (and huge font files) to support so this is not an option for us. I couldn't successfully embed that many fonts into my app. Still eagerly awaiting the fix.

**Comment by ajwfrost** [comment]
Hi

Yes I assumed that you'd need to use the device fonts for a reason ... well we have a new mechanism for text rendering which uses the "normal" Android drawing routines. There are some issues still with positioning/scrolling and transparency that we're working through but it shouldn't take too much longer - I'll try to get you a preview version asap..

thanks

**Comment by yutfa** [comment]
> I have the same problem, but I solved it temporarily...
> 
> My players reported this bug to me, too and their phones are all Pixel series + android 12 finally, I guess ....it must lack certain a font...
> 
> Then, I embedded a Font in my game...and use it in TextField(use .embedFonts = true) all the problem is solved...
> 
> if you guys are in a hurry, you could try this method first
@mucoci 

Embedding fonts can only solve normal text, input text is still garbled

**Comment by mucoci** [comment]
> > I have the same problem, but I solved it temporarily...
> > My players reported this bug to me, too and their phones are all Pixel series + android 12 finally, I guess ....it must lack certain a font...
> > Then, I embedded a Font in my game...and use it in TextField(use .embedFonts = true) all the problem is solved...
> > if you guys are in a hurry, you could try this method first
> > @mucoci
> 
> Embedding fonts can only solve normal text, input text is still garbled

so... it is not a perfect solution...
just a temporary one...unless your input text only including a-z, A-Z or other symbol

**Comment by a9848840** [comment]
Hi
@ajwfrost 

I use Air 33.1.1.743
But the problem is not solved
I hope you will continue to finish
thanks

**Comment by matrix3d** [comment]
fast!

**Comment by ajwfrost** [comment]
Yes we're still looking at it - struggling with the mixture of drawing in Java and then compositing onto a surface in C++ that may actually be an Android bitmap or an OpenGL texture. The earlier code just short-cutted all of that but used private non-NDK code and libraries that are then not working properly with Android 12 hence the issue.

Currently (with our work in progress), text drawn with a background colour will be fine. It's just blending this onto any existing background that's causing a problem..

**Comment by Devin-Licastro** [comment]
I would feel a bit better if you had said "this is our top priority" rather than "we are still looking at it". 

Our app is a b2b corporate training platform and we're working with clients in China, HongKong, Taiwan, Korea, and Japan that are all heavily impacted by this issue. Every day more and more phones are launching android 12 updates and more and more of our users can't use the app anymore. 

Our clients are freaking out that they paid us for this service that we can no longer provide. We've already had to take a large financial hit as we've needed to offer discounts to affected clients, not to mention the potential churn we will face during the next contract renewal periods. These discounts and our attempts to reassure them a fix is coming while begging them for more time won't hold them off much longer. This is an absolute disaster for us and could mean the end of our business entirely if it isn't handled soon. 

Please, please, put all of your resources into this. What else could be more important than a critical issue that renders a significant percentage of your userbase unable to use Air at all anymore?

**Comment by ajwfrost** [comment]
Thanks for the perspective/impact .. I've had someone working on it full time but I think we suffered from an initial misdirection and then more recently a problem where the existing behaviour wasn't properly understood. We're at a point now where the cpu-mode rendering is working and we're trying to resolve problems with the direct-mode rendering.

I think it may be worth us also sending you a test file so that you can double-check the fix (and so you can push it out if you find it works) - can you confirm if you do use cpu renderMode, or do you need direct mode?

thanks

**Comment by Devin-Licastro** [comment]
Our project is starling/feathers based so we need the direct mode, unfortunately.

**Comment by ajwfrost** [comment]
Hi @Devin-Licastro, @matrix3d, @a9848840  - if you're able to double-check what happens if you use the below runtime build (for armv8, on top of the latest SDK or anything above 698 I think), it would be good to get some "real" feedback!

https://transfer.harman.com/message/YOGggmbg3i7vblG8VMUfTl

thanks

**Comment by matrix3d** [comment]
can you make three new api.
var bmd=Font.getBmd("abc");
Font.drawToBmd();
Font.drawToTexture();

**Comment by a9848840** [comment]
Hi @ajwfrost 

I use Air 33.1.1.743
https://transfer.harman.com/message/YOGggmbg3i7vblG8VMUfTl 
change Runtime.apk
crash log

01-25 10:54:00.589   575   585 D libnos_transport: Calling App 2 with params 0x0000
01-25 10:54:00.590   575   585 E libnos_datagram: can't send spi message: Try again
01-25 10:54:00.595   575   585 E libnos_datagram: can't send spi message: Try again
01-25 10:54:00.601   575   585 E libnos_datagram: can't send spi message: Try again
01-25 10:54:00.606   575   585 E libnos_datagram: can't send spi message: Try again
01-25 10:54:00.612   575   585 E libnos_datagram: can't send spi message: Try again
01-25 10:54:00.617   575   585 E libnos_datagram: can't send spi message: Try again
01-25 10:54:00.625   575   585 D libnos_transport: App 2 inspection status=0x00000000 reply_len=0 protocol=1 flags=0x0000
01-25 10:54:00.625   575   585 D libnos_transport: Send app 2 command data (66 bytes)
01-25 10:54:00.626   575   585 D libnos_transport: Send app 2 go command 0x00020000
01-25 10:54:00.626   575   585 D libnos_transport: Polling app 2
01-25 10:54:00.659   575   585 D libnos_transport: App 2 poll=16 status=0x80000000 reply_len=0 flags=0x0000
01-25 10:54:00.660   575   585 D libnos_transport: App 2 polled=16 status=0x80000000 reply_len=0 flags=0x0000
01-25 10:54:00.660   575   585 D libnos_transport: App 2 returning 0x0
01-25 10:54:00.774  1704  1937 I system_server: Explicit concurrent copying GC freed 1846195(70MB) AllocSpace objects, 111(5192KB) LOS objects, 71% free, 38MB/134MB, paused 655us,67us total 180.136ms
01-25 10:54:00.782   984  2459 I sensors-hal: [ssc_utils]get_qmi_debug_flag:243, support_qmi_debug : false
01-25 10:54:00.794  1704  1937 I ActivityManager: Force stopping PackageName appid=10650 user=0: pkg removed
01-25 10:54:00.815   984  2459 I sensors-hal: [ssc_utils]get_qmi_debug_flag:243, support_qmi_debug : false
01-25 10:54:00.844  1704  1721 W System  : A resource failed to call close.
01-25 10:54:00.844  1704  1721 W System  : A resource failed to call close.
01-25 10:54:00.844  1704  1721 W System  : A resource failed to call close.
01-25 10:54:00.846  1704  1721 W System  : A resource failed to call release.
01-25 10:54:00.847  1704  1721 W System  : A resource failed to call close.
01-25 10:54:00.849  1704  1721 W System  : A resource failed to call close.
01-25 10:54:00.870  1704  1721 W System  : A resource failed to call close.
01-25 10:54:00.872  1704  1721 W System  : A resource failed to call release.
01-25 10:54:00.872  1704  1721 W System  : A resource failed to call close.
01-25 10:54:00.873  1704  1721 W System  : A resource failed to call close.
01-25 10:54:00.890  9354  9354 E AppButtonsPrefCtl: Exception when retrieving package:PackageName
01-25 10:54:00.890  9354  9354 E AppButtonsPrefCtl: android.content.pm.PackageManager$NameNotFoundException: PackageName
01-25 10:54:00.890  9354  9354 E AppButtonsPrefCtl:     at android.app.ApplicationPackageManager.getPackageInfoAsUser(ApplicationPackageManager.java:241)
01-25 10:54:00.890  9354  9354 E AppButtonsPrefCtl:     at android.app.ApplicationPackageManager.getPackageInfo(ApplicationPackageManager.java:213)
01-25 10:54:00.890  9354  9354 E AppButtonsPrefCtl:     at com.android.settings.applications.appinfo.AppButtonsPreferenceController.retrieveAppEntry(AppButtonsPreferenceController.java:368)
01-25 10:54:00.890  9354  9354 E AppButtonsPrefCtl:     at
[android12.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/7930318/android12.txt)
 com.android.settings.applications.appinfo.AppButtonsPreferenceController.refreshUi(AppButtonsPreferenceController.java:656)
01-25 10:54:00.890  9354  9354 E AppButtonsPrefCtl:     at com.android.settings.applications.appinfo.AppButtonsPreferenceController.refreshAndFinishIfPossible(AppButtonsPreferenceController.java:504)

**Comment by matrix3d** [comment]
we have code the embedFonts.it is works fine

**Comment by matrix3d** [comment]
and the stagetext also worked

**Comment by a9848840** [comment]
Because my app is not small
Lots of places with text fields
If it can be solved once
That's the best result

**Comment by matrix3d** [comment]
我昨天用几个小时就解决了，动态加载字体swf。多了5m的大小。我们程序只需要改2个地方就可以了。

**Comment by matrix3d** [comment]
网络上的动态加载字体代码都是错的。昨天自己研究了一套方案

**Comment by ajwfrost** [comment]
@a9848840 interesting log .. thanks for that. The failure is a null pointer when calling memcpy - the destination address is zero. Looking at the register values and decompiling the code, it seems we've got a zero-width bitmap being passed in for some reason, which we're not properly handling.

Just building an updated version and will post it here as soon as it's ready..

thanks

**Comment by ajwfrost** [comment]
@matrix3d 
> can you make three new api.
> var bmd=Font.getBmd("abc");
> Font.drawToBmd();
> Font.drawToTexture();

Can you clarify the suggestion here? The first one seems like you want a bitmap data from passing a text string to a Font object but there's no setting of size/colour/etc in there? Not sure about the drawToBmd/drawToTexture options, the first one presumably would be the same as getBmd?

If you're looking for a new/utility API to help with the generation of bitmapdata or texture objects based upon text strings and fonts, can you please raise a new issue and tag it with 'feature'?

thanks

**Comment by matrix3d** [comment]
https://github.com/airsdk/Adobe-Runtime-Support/issues/1597

**Comment by ajwfrost** [comment]
Thanks

@a9848840 please re-try with this one:
https://transfer.harman.com/message/hLsy3H0EPZe8SsUuZJXi8A

thanks

**Comment by a9848840** [comment]
Hi

@ajwfrost 
charh log
[android12_1.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/7932994/android12_1.txt)

The device is not Android 12, but the crash still occurs
The time when the crash occurred, which is slower than the last Runtime

**Comment by ajwfrost** [comment]
Thanks @a9848840  .. that's a curious one.
Is it possible to get an apk-debug build of your app so that we can test it here? (plus any repro steps)?
https://transfer.harman.com/requests/UOoiRrt3Y4QS1u2MKC3Gsg
thanks

**Comment by ajwfrost** [comment]
Actually it looks like o.o.m. and then we didn't handle the JNI issue...
We can try to get a reproduction of this - not sure whether we need to call 'recycle' on the bitmap, there is a canvas that has a reference to it but the canvas is no longer referenced so it should work through GC anyway...

```
01-25 19:02:58.528 22033 22033 W System.err: java.lang.OutOfMemoryError: Failed to allocate a 12582924 byte allocation with 9533408 free bytes and 9MB until OOM
01-25 19:02:58.528 22033 22033 W System.err:    at dalvik.system.VMRuntime.newNonMovableArray(Native Method)
01-25 19:02:58.528 22033 22033 W System.err:    at android.graphics.Bitmap.nativeCreate(Native Method)
01-25 19:02:58.528 22033 22033 W System.err:    at android.graphics.Bitmap.createBitmap(Bitmap.java:895)
01-25 19:02:58.528 22033 22033 W System.err:    at android.graphics.Bitmap.createBitmap(Bitmap.java:872)
01-25 19:02:58.528 22033 22033 W System.err:    at android.graphics.Bitmap.createBitmap(Bitmap.java:839)
01-25 19:02:58.528 22033 22033 W System.err:    at com.adobe.air.customHandler.callTimeoutFunction(Native Method)
01-25 19:02:58.528 22033 22033 W System.err:    at com.adobe.air.customHandler.handleMessage(customHandler.java:28)
01-25 19:02:58.528 22033 22033 W System.err:    at android.os.Handler.dispatchMessage(Handler.java:102)
01-25 19:02:58.528 22033 22033 W System.err:    at android.os.Looper.loop(Looper.java:159)
01-25 19:02:58.528 22033 22033 W System.err:    at android.app.ActivityThread.main(ActivityThread.java:6130)
01-25 19:02:58.528 22033 22033 W System.err:    at java.lang.reflect.Method.invoke(Native Method)
01-25 19:02:58.528 22033 22033 W System.err:    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
01-25 19:02:58.528 22033 22033 W System.err:    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
01-25 19:02:58.559 22033 22033 F art     : art/runtime/java_vm_ext.cc:470] JNI DETECTED ERROR IN APPLICATION: obj == null
01-25 19:02:58.559 22033 22033 F art     : art/runtime/java_vm_ext.cc:470]     in call to CallIntMethodA
```

**Comment by Devin-Licastro** [comment]
"please re-try with this one:"
Hey, so I tried my app with newest air SDK and this runtime and it was quite promising! 

Couple of notes:

- For our app we embed an English font (but we don't embed fonts for other languages), also note that the user can choose a language and the UI will appear in that language
- So when app UI language was set to English, the textfields attempting to use the embedded English fonts were no longer rendered
- But when setting the app to another language, (Chinese, Japanense, etc) all the text fields (attempting to use device font) were now successfully rendered... no more [], and I didn't notice any graphical issues or anything like that
- For android 12 testing, I don't have an android 12 device yet myself, but I use AWS device farm to run apk's on their android 12 devices. (I tested on the pixel 6 pro.) So I used this method to confirm that the Chinese font is rendering now. But I can't comment on the app performance as the app always appears to run really slowly on the AWS device farm. 
- I'm not super knowledgeable about the inner workings of textfields for starling / feathers, but half the places in the app use default starling textfields, and other half use like the 'label' field of a feathers button for example. We have our feathers default text render factory set as:

FeathersControl.defaultTextRendererFactory = function():ITextRenderer
{
	var renderer:TextFieldTextRenderer = new TextFieldTextRenderer();
	return renderer;
};

**Comment by ajwfrost** [comment]
Thanks for the feedback...

> the textfields attempting to use the embedded English fonts were no longer rendered

I wouldn't have expected anything we did to change how the embedded font rendering worked .... but we will check on this. One of the team has been doing some further testing, I should get an update tomorrow.

thanks

**Comment by ajwfrost** [comment]
@Devin-Licastro  having just tested this myself quickly, I can't see a problem with embedded fonts; are you able to provide any code snippets or samples that demonstrate the issue?

Have tried both the TextField and TextBlock/TextLine mechanisms, cpu and direct mode rendering, but I can see the text appearing as normal in both cases..

thanks

**Comment by Devin-Licastro** [comment]
Hmm, we embed the fonts like this:

`
[Embed(source="../../../../assets/fonts/proxima-nova-regular.ttf", embedAsCFF="false", fontFamily="Prox_R", fontWeight="normal")]
private static const customFont:Class;
`
		
Then in general we create textfields like:
`
var textFormat:TextFormat = new TextFormat('Prox_R', 24, 0xffffff); 
var tf:TextField = new TextField(200, 50, "Text to display");
tf.format.setTo(textFormat.font, textFormat.size, 0xfffff); 	
tf.pixelSnapping = true;
`

Let me know if this doesn't help and a sample project is needed

**Comment by ajwfrost** [comment]
Okay thanks, let me get a test together using Starling for this...

**Comment by a9848840** [comment]
Hi
@ajwfrost 

I need to inform my settings additionally

xml
<BuildLegacyAPK>true</BuildLegacyAPK>

jdk1.8.0_111

**Comment by Devin-Licastro** [comment]
Awesome, I've confirmed the fix has worked, great to see!
Looking forward to the release!

About the request for logcat filtered by AdobeAIR, not much to see, just the loading of ANEs and some creation/deletion of audio tracks. Here's how it looks:

`C:\Air\air33.1.1.743\lib\android\bin\adb.exe logcat AdobeAIR:V *:S
adb server version (32) doesn't match this client (41); killing...
* daemon started successfully
--------- beginning of crash
--------- beginning of perf
--------- beginning of system
--------- beginning of main
01-26 18:54:16.928 12102 12102 I AdobeAIR: Loading extension androidx.core on Android-ARM - SWF data 0xe3dc1850
01-26 18:54:16.932 12102 12102 I AdobeAIR: Loading extension androidx.appcompat on Android-ARM - SWF data 0xe3dc1850
01-26 18:54:16.935 12102 12102 I AdobeAIR: Loading extension androidx.vectordrawable on Android-ARM - SWF data 0xe3dc1850
01-26 18:54:16.938 12102 12102 I AdobeAIR: Loading extension androidx.browser on Android-ARM - SWF data 0xe3dc1850
01-26 18:54:16.941 12102 12102 I AdobeAIR: Loading extension androidx.exifinterface on Android-ARM - SWF data 0xe3dc185001-26 18:54:16.944 12102 12102 I AdobeAIR: Loading extension com.distriqt.square.okhttp3 on Android-ARM - SWF data 0xe3dc1850
01-26 18:54:16.948 12102 12102 I AdobeAIR: Loading extension com.distriqt.square.picasso on Android-ARM - SWF data 0xe3dc1850
01-26 18:54:16.952 12102 12102 I AdobeAIR: Loading extension com.distriqt.Share on Android-ARM - SWF data 0xee14db70
01-26 18:54:16.956 12102 12102 I AdobeAIR: Loading extension com.distriqt.NativeWebView on Android-ARM - SWF data 0xe30c4330
01-26 18:54:16.962 12102 12102 I AdobeAIR: Loading extension com.distriqt.Core on Android-ARM - SWF data 0xe3a03260
01-26 18:54:16.966 12102 12102 I AdobeAIR: Loading extension com.distriqt.Dialog on Android-ARM - SWF data 0xe03aa0f0
01-26 18:54:16.970 12102 12102 I AdobeAIR: Loading extension com.distriqt.PDFReader on Android-ARM - SWF data 0xcdd814a001-26 18:54:16.974 12102 12102 I AdobeAIR: Loading extension com.distriqt.MediaPlayer on Android-ARM - SWF data 0xdb480010
01-26 18:54:16.979 12102 12102 I AdobeAIR: Loading extension com.distriqt.Flurry on Android-ARM - SWF data 0xe0315010
01-26 18:54:16.983 12102 12102 I AdobeAIR: Loading extension com.distriqt.playservices.Base on Android-ARM - SWF data 0xdb440010
01-26 18:54:16.986 12102 12102 I AdobeAIR: Loading extension com.distriqt.playservices.Analytics on Android-ARM - SWF data 0xe3dc1850
01-26 18:54:16.990 12102 12102 I AdobeAIR: Loading extension com.distriqt.playservices.AdsIdentifier on Android-ARM - SWF data 0xe3dc1850
01-26 18:54:16.993 12102 12102 I AdobeAIR: Loading extension com.distriqt.CustomResources on Android-ARM - SWF data 0xe03c1230
01-26 18:54:16.996 12102 12102 I AdobeAIR: Loading extension com.distriqt.Notifications on Android-ARM - SWF data 0xe087c1f0
01-26 18:54:17.000 12102 12102 I AdobeAIR: Loading extension com.distriqt.Application on Android-ARM - SWF data 0xdb480010
01-26 18:54:17.006 12102 12102 I AdobeAIR: Loading extension com.distriqt.Scanner on Android-ARM - SWF data 0xcdd814a0
01-26 18:54:17.010 12102 12102 I AdobeAIR: Loading extension com.distriqt.CameraRollExtended on Android-ARM - SWF data 0xebf84030
01-26 18:54:17.014 12102 12102 I AdobeAIR: Loading extension com.distriqt.CameraUI on Android-ARM - SWF data 0xe0315010
01-26 18:54:17.018 12102 12102 I AdobeAIR: Loading extension com.distriqt.NetworkInfo on Android-ARM - SWF data 0xdb2c9cf0
01-26 18:54:17.021 12102 12102 I AdobeAIR: Loading extension com.distriqt.PushNotifications on Android-ARM - SWF data 0xe30c4330
01-26 18:54:17.026 12102 12102 I AdobeAIR: Loading extension com.distriqt.SystemGestures on Android-ARM - SWF data 0xcf007230
01-26 18:54:17.462 12102 12102 D AdobeAIR: Creating audio track (mixer 0xebf9a0e0) from thread 12102
01-26 18:54:29.780 12102 12102 D AdobeAIR: Destroying audio track (mixer 0xebf9a0e0) from thread 12102
01-26 18:55:03.403 12102 12102 D AdobeAIR: Creating audio track (mixer 0xebf9a0e0) from thread 12102
01-26 18:55:18.171 12102 12102 D AdobeAIR: Destroying audio track (mixer 0xebf9a0e0) from thread 12102`

**Comment by wmtj** [comment]
AIRSDK 33.1.1.758,  Chinese characters worked, but the performance was terrible. On some of the older android phones, the lag was obvious and almost unusable. So to say, This bug,  remains unsolved.

**Comment by ajwfrost** [comment]
@wmtj can I check on the older phones, was the previous mechanism working okay still in earlier AIR releases? The 'new' mechanism is using the standard Android rendering for fonts which is slower as we have to use JNI, although we've been reviewing this and think we can see an optimization to avoid some new bitmap creation that would currently be happening in 'direct' mode SWFs. But it might be that we can continue to use the old approach if the running Android version is < 12..

Just to double check: were the apps on the older phones using 'direct' or 'cpu' mode for rendering?

thanks

**Comment by wmtj** [comment]
> 'direct' mode.

**Comment by wmtj** [comment]
Maybe, “to use the old approach if the running Android version is < 12” is not a good choice, Unless Harman will fix those old Chinese font display bugs that have been around for years (for example: on a part of phones with android8/10/11(not all), '_sans' font does not work, it shows just like '_serif'. the '_serif' makes some chinese characters dispaly in wrong form,such as '才','门','关'...). @ajwfrost

**Comment by wmtj** [comment]
As far as I can remember, the bugs I mentioned above, Adobe had fixed it in Android 5 era, but it repeated in Android 8 and later. At that time, most of AIR develpers agreed that the AIR was dying, no one would paying attention to this problem, and no one in Adobe would fix it. Fortunately, Harman took over the the AIR development, I saw the hope again.But unfortunately, Harman seems never to have tested Chinese before every AIRSDK releases.It was a huge mistake, especially after the release of Android 12 Beta.
For now, I just hope Harman will be able to make a thorough argument before making a technical decision, so as not to solve one issue and create more issues.

If you have no one knows Chinese in your team, please listen to your users and we can provide help if necessary.I am sorry for my poor English, I hope my meaning is clear :) 
@ajwfrost

**Comment by matrix3d** [comment]
门 为啥会显示错误，有截图吗，正确的啥样，错误的啥样

**Comment by ajwfrost** [comment]
@wmtj  the reason for all these issues with the text is because of how AIR did the font rendering. In earlier AIR versions (i.e. everything, up to our last beta 33.1.1.1.758) there is a build of the Skia library which is what Android use for their rendering, but this is slightly customized by Adobe; the problem seems to be that in order to optimize the rendering, it loses the wider functionality that Android has for font fallback handling. There are a few workarounds that I can see have been added, but this then broke in the latest Android release.

Our choices here were then:
1) try to continue patching and working around the latest Android changes using the older Skia library with the native rendering methods; or
2) switch AIR to render text in the same way that Android apps do, which should then improve the functionality and is more future-proof.

We did start with the first option but part of the challenge is to do this, we need a build environment where we can build the whole Android source code from 4.0. And that's just more and more of a problem as the tools and environments no longer exist. Hence the idea of looking at the new mechanism....  although there have obviously been a few challenges with this, in particular for GPU mode and because of the performance (having to use JNI and also some copying of pixels from 'Android/Java' structures into the AIR framebuffer).

So we're dealing with some of the limitations of Android and the fact that the NDK doesn't give us all of the abilities that we need in order to optimize the rendering. But I think with the fixed/optimizations that we're working on currently, it should improve things a fair bit, and we'll see if there are additional updates we can make in terms of caching text images if possible.

We don't have anyone with native Chinese on our immediate team but we very much appreciate the inputs and comments that folk are making here, I have to confess I wasn't aware that there were such historical issues with Chinese text in AIR in those earlier Android versions.

We'll get a new beta build out before too long with the latest optimizations so hopefully this should improve it.

**Comment by wmtj** [comment]
@matrix3d  @ajwfrost 
[Compare](https://www.ziwei001.com/update/beta/test/compare.jpg) 

This issue only occurs on a part of phones with android 8/10/11, not all.  It might be fine before a system upgrade, but not fine after the upgrade. I guess the customized Android from various manufacturers may change some of the font files in specific 
 versions , the AIRSDK detected it and automatically change '_sans' to '_serif' or others. As the picture above shows, the font is very weird.

**Comment by matrix3d** [comment]
@wmtj 这是什么原因。为啥门字，会变成错误的。我在游戏中也发现了这个问题

**Comment by matrix3d** [comment]
也就是那个错误的门是啥玩意？日本字？

**Comment by matrix3d** [comment]
@ajwfrost what diff textfield and stagetext,the stagetext.drawtobitmap work good

**Comment by ajwfrost** [comment]
Hi @matrix3d 

The StageText implementation is an Android text field (`android.widget.EditText`) that's displayed in an overlay layer on the screen, so all the calls that you make to format etc in ActionScript are sent through C++ into the Java layer to change that text field. When you capture it in a bitmapdata object, it's then creating an Android `Bitmap` object and a `Canvas` associated with this, drawing the StageText container into this canvas, and then getting the pixels from that Bimap.

So that's not massively efficient, but it's fairly similar now to what we're doing. The challenge that we've faced is to integrate into the rendering mechanisms .. there are different approaches for the various different render modes in AIR, hence e.g. not working in GPU mode as they are all set up as textures without background framebuffer objects; in direct mode it seems we've got some issue with elements disappearing which may be due to clip rectangles or caused by the copying ... some of the inefficiencies are because we've got a canvas/bitmap for the text rendering, but this is being drawn on top of something else so we need to allow it to draw what's underneath, then grab that part of the frame buffer and get Android to draw the text on top, and then copy that back into the main frame buffer for anything else to go above that.

Anyway -> getting some progress, I'll try to get an updated runtime posted here later on so that you can all check it...

thanks

**Comment by ajwfrost** [comment]
FYI there's an updated runtime for arm64 available below. Needs some further work and optimizations as well but would value any feedback!
https://github.com/airsdk/Adobe-Runtime-Support/issues/1630#issuecomment-1037230015

thanks

**Comment by wmtj** [comment]
@ajwfrost 
1. 'Textfield.cacheAsBitmap=true' cause Buttons and other DisplayObjects disappeared.
2. 'Textfield.useEmbedFont=true' cause textfields disappeared. 31.1.1.758 is OK.
3. Textfield.width & Textfield.textWidth, it still looks wrong.
4. The last character is still cut, same to 31.1.1.758.
5. Textfield colors are totally messed up.

**Comment by wmtj** [comment]
@ajwfrost 
The performance is a little better than previous verison, but it is still too slow compared to 31.1.1.6xx. In my App, there is a complex page with more than 200 textfields. I tested it on an old phone (Kirin 659 Processor, Android 9), the old version app with AIRSDK 31.1.1.6xx takes only 2 seconds to show, while the new verison with the runtime you given above takes 34 seconds(First time execute after launched), 8 seconds(execute again)    :(

**Comment by wmtj** [comment]
@ajwfrost 
This bug has been going on for almost 4 months, and we have lost a lot of users.If you are still unable to improve AIRSDK performance to the previous level in February, I suggest you consider“to use the old approach if the running Android version is < 12” as a temporary solution, as soon as quickly! Thus, at least we can release a new version to the Play Store.

**Comment by matrix3d** [comment]
@wmtj 用嵌入字体

**Comment by wmtj** [comment]
@matrix3d 
嵌入字体可以解决一般场景的汉字显示，比如菜单、按钮等的显示都是没有问题的，但涉及到复杂的场景，则有局限性,例如：

1.一系列 TextFields 先被添加到Sprite上，这个Sprite的尺寸不是固定不变的，要适应不同尺寸的屏幕或者应对某些伸缩变化，改变Sprite的width和height，如果使用嵌入式字体，则会造成字体纵向拉长或者横向压扁。而使用AIR的“_sans”则没有字体拉伸问题。 

2.某些需要中文、日文、韩文和泰文等东亚语言文字两种或者多种混合显示的场景，使用嵌入式字体往往力不从心：嵌入中文字体，TextField指定该字体名后则只能显示中文，显示不了泰文……而使用AIR的“_sans”则没有此限制。

以上，也有可能是我的使用方法不对，如有错误之处，请大家不吝指出。谢谢！


Embedded fonts can solve the problem of displaying Chinese characters in general scenes, such as menus and buttons, but there are limitations when it comes to complex scenes, such as:

1. A series of TextFields are added to a Sprite. The size of the Sprite is not fixed.Changing the width and height of the Sprite to accommodate different screen sizes or to accommodate some scaling changes can cause the font to stretch lengthwise or flatten horizontally if you use an embedded font.AIR "_sans"  has no font stretching problems.

2. In some scenes where two or more east Asian languages such as Chinese, Japanese, Korean and Thai are required to be displayed together, it is often difficult to use an embedded font: If embedded Chinese font, when TextField specifies the font name, it can only display Chinese, but not Thai. AIR "_sans" has no this limitation.

Above, if there are mistakes, please tell me. Thank you!

**Comment by ajwfrost** [comment]
@wmtj I think the first of those issues - with stretched text - should not happen with embedded fonts unless something's been set up slightly wrongly or unless they're being rendered to a bitmap..?

Anyway -> we are going with your suggestion from earlier which means a version that uses the old library for pre-Android 12 devices and the new mechanism for 12+, with a view to then (a) optimizing the new code, or (b) trying to get a separate native library together from a more modern Android build that we can use without needing JNI...

thanks

**Comment by wmtj** [comment]
@ajwfrost 
Here is an example of 'EmbedFont cause stretched text' : 
[Screenshot](https://www.ziwei001.com/update/beta/test/Screenshot.png)
[Code](https://www.ziwei001.com/update/beta/test/src.zip)

The example shows the difference between TextField with '_sans' and embedFonts while scaling their parent Object height(keep width the same) : The width and height of TextField with '_sans' are enlarged in equal proportion, TextField with embedFonts enlarged height observably, width is enlarged slightly. 

Is there anything wrong in the example code ?

**Comment by ajwfrost** [comment]
Thanks @wmtj  .. so I see what you mean. And I can kind of see it two ways here! What's essentially happening is a sprite is having a text field drawn onto it with a size of 1000x150, and then the sprite height is changed to 220 whilst the width stays the same. So in a way you could assume this is intended to stretch the contents of the sprite - which is actually what the documentation says too (for DisplayObject.height, when you set this it changes the scaleY property...).

So I suspect we could almost say that it's the device text (non-embedded fonts) that's not behaving properly, although of course device text is drawn by the device and doesn't support scaling etc which is why we've ended up with this discrepancy.

Seems a bit weird but I'm not sure we should change how this works, it's very likely that there are people who depend on this behaviour!

An interesting aside though, thanks for the example...

**Comment by wmtj** [comment]
@matrix3d  确实如你所说，这两种方式有不同的表现和用途，在我的使用场景里无法用embedFont代替'_sans'，这就是我为何急于请ajwfrost尽快发布可用版本的原因。

@ajwfrost Yes, It is an interesting example that explains why I can't use  embedFont to solve the issue.

**Comment by FliplineStudios** [comment]
> Anyway -> we are going with your suggestion from earlier which means a version that uses the old library for pre-Android 12 devices and the new mechanism for 12+, with a view to then (a) optimizing the new code, or (b) trying to get a separate native library together from a more modern Android build that we can use without needing JNI...

@ajfrost Since the classic text rendering will still be present and usable in the SDK as you mentioned, is there any chance we can opt-in/opt-out of this new rendering mechanism for Android 12+, perhaps with a setting in our app xml?  For developers like us that don't ever use Chinese text or emoji in their apps, I'd love to be able to have the same classic text-rendering experience in our apps regardless of Android version, and not have to worry about performance or other issues arising from some of our end users running Android 12.

**Comment by mucoci** [comment]
> 

It's not a good idea~

Actually, the garbled text is not only in Chinese.. also in Japanese, Korean......
All the characters like this type will be garbled...

Unless your app only faces the languages which use alphabetic like English and French...
Otherwise, you will need to handle this in the future

**Comment by FliplineStudios** [comment]
Yes that's what I meant, our apps only use the Latin alphabet and limit the characters for text input, so while this new text rendering is still experimental it would be great for developers to opt-in/opt-out so they can use the classic rendering in the meantime, and not have to stick with an older AIR SDK version until things are stable.

I was thinking more along the lines of back when Adobe made a big change to their AOT compiler, and we could specify the adt -useLegacyAOT flag while they were still fine-tuning it to opt out and stick with the original AOT compilation temporarily, and then eventually when it was considered stable the useLegacyAOT flag was ignored during compilation.

**Comment by a9848840** [comment]
Hi
@ajwfrost 

I'm sorry to keep urging you, because android 12 text is related to the company's revenue, can you focus on the text part first?

thank you

**Comment by ajwfrost** [comment]
Hi @a9848840  - I think we've got a much better implementation now, at least from the test cases that we've got here, and it's just going through the last QA processes this morning. All being well, we'll push it up to the website tonight.

thanks

**Comment by ajwfrost** [comment]
Hi @a9848840, @wmtj  please try the latest build of the AIR SDK:
https://airsdk.harman.com/download/33.1.1.779

thanks

**Comment by a9848840** [comment]
Hi
@ajwfrost 

Initial installation and use test
functioning normally
Thank you very much

**Comment by wmtj** [comment]
@ajwfrost Thank you and your team for the hard working!

The performance is much better than the previous 758. It is acceptable on Android 12 now. 
It is still slow on old android phone with new rendering mechanism. After setting '<newFontRenderingFromAPI>31</newFontRenderingFromAPI>',it works.

I found another severe issue(At least it seems to be): A white screen may occur after a specific operation in 'direct' mode. In 'cpu' mode, it works fine, no white screen. I have recorded the 'white screen' video and will send the link to your email later. My app is not a game app, it can run both 'direct' and 'cpu' mode, but in 'cpu' mode, the display effect of Chinese characters is significantly inferior to 'direct' mode with old rendering mechanism.

Although it still doesn't look perfect, at least it works. Maybe I should pack an armv7 package using 33.1.1.743, armv8 using 33.1.1.779 .

**Comment by a9848840** [comment]
HI
@ajwfrost @marchbold 

AIR SDK 33.1.1.779

A crash occurs on ADMOB operation
It happens 1 time out of 5 when I test
At the trigger point of requesting an ad

Should I open another question?
[crash.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/8116716/crash.txt)

**Comment by ajwfrost** [comment]
@wmtj - are you able to create a demo showing the cut characters? not sure what's happening with that..
and in the white screen, do you see anything on the logcat output?

@a9848840  thanks for the log - nothing to do with admob from what I can see:
```
02-22 17:16:49.019 32087  1569 F winner.bwmj16t: java_vm_ext.cc:675] JNI ERROR (app bug): global reference table overflow 02-22 17:16:49.019 32087  1569 F winner.bwmj16t: java_vm_ext.cc:675]   Summary:
02-22 17:16:49.019 32087  1569 F winner.bwmj16t: java_vm_ext.cc:675]     49626 of android.graphics.Paint (49626 unique instances)
```
So this means we're creating Paint objects but not destroying them .. will have to work out where this happens!

thanks

**Comment by wmtj** [comment]
@ajwfrost   I've spent the whole day trying to build a demo that recreates the white-screen problem, but I failed.  I do not know the cause of the problem and how to trigger mechanically, so …… I  will send the link to your email later, I hope you can find it in the phenomenon.

**Comment by wmtj** [comment]
@ajwfrost  Here is a demo showing the cut characters. It looks like that if the first character is a Chinese punctuation mark, TextField.textWidth will be wrong value.This is a pattern that I have just find, and there are many more that have yet been unfound.

[ScreenShot](https://www.ziwei001.com/update/beta/test/ScreenShot3.jpg)

Code  ：
`var _tfmt1:TextFormat = new TextFormat("_sans", 50, 0);
_tfmt1.align = TextFormatAlign.CENTER;

var _txt1:TextField = new TextField();
_txt1.defaultTextFormat = _tfmt1;
_txt1.text = "简体中文测试，繁體中文測試。";
_txt1.width = _txt1.textWidth + 12;
_txt1.height = _txt1.textHeight + 4;
_txt1.border = true;
_txt1.x = 100;
_txt1.y = 200;
addChild(_txt1);

var _txt2:TextField = new TextField();
_txt2.defaultTextFormat = _tfmt1;
_txt2.text = "简体中文测试（繁體中文測試）测试";
_txt2.width = _txt2.textWidth + 12;
_txt2.height = _txt2.textHeight + 4;
_txt2.border = true;
_txt2.x = 100;
_txt2.y = 300;
addChild(_txt2);

var _txt3:TextField = new TextField();
_txt3.defaultTextFormat = _tfmt1;
_txt3.text = "（简体中文测试繁體中文測試）测试";
_txt3.width = _txt3.textWidth + 12;
_txt3.height = _txt3.textHeight + 4;
_txt3.border = true;
_txt3.x = 100;
_txt3.y = 400;
addChild(_txt3);`

**Comment by a9848840** [comment]
Thank you for your team's hard work, please keep it up, thank you

**Comment by ajwfrost** [comment]
@wmtj thanks for that .. so it looks like the "bounds" that we get back from Android (`Paint.getTextBounds()`) will be just a bounding rectangle that encloses the pixels that would change, rather than including the space at the start of that "（" character. If we switch to using `Paint.measureText()` then it gives us an accurate value for the width of the text as we'd need when appending it onto other pixels...  So, with this change, I can see that test case fixed...

I didn't get any email with the video on it, it may have been rejected/blocked by our email systems. In case you can upload something, please try using this link: https://transfer.harman.com/requests/vYK8tct7lGKQUaqGZhnID1

@a9848840 we did have our test case for this crash with the same error after we left it for a few hours. Once we changed to ensure the specialized class's destructor was properly run, the same test case has just been running overnight and is still going 10 hours later.

thanks

**Comment by a9848840** [comment]
Hi
@ajwfrost 

I currently have a situation using AIR SDK: 33.1.1.779

After package apk file and re-sign
upload file to googleplay store
64-bit re-signed apk received the following message

The APK you uploaded is not compressed and calibrated. Please run the compressed calibration tool on your APK and upload it again.

32 bit apk update normally

Then I upload the unresigned 64-bit file
can be uploaded normally

Do I need to make changes to re-sign 64-bit?
[Resign64.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/8130296/Resign64.txt)

**Comment by ajwfrost** [comment]
Hi @a9848840  - am not sure on the process you're using there, did you do anything to the 64-bit APK between the first attempt which failed and the second attempt (after the 32-bit upload) which worked? Not sure why you're needing to re-sign it...?

**Comment by a9848840** [comment]
hi, @ajwfrost 

Because it is used in the APP
AppsFlyerAIRExtension-witout-gp-support.ane
AppsFlyerAIRExtension.swc
need to re-sign

airsdk:33.1.1.554 re-sign did not happen
airsdk:33.1.1.779 re-sign the above happens

i try to update
https://repo.maven.apache.org/maven2/com/appsflyer/af-android-sdk/6.5.2/af-android-sdk-6.5.2.aar
AppData\Local\Android\Sdk\build-tools\32.0.0\zipalign.exe
Still unable to upload successfully
looking for other ways

[Resign64.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/8131357/Resign64.txt)

**Comment by Ender22** [comment]
@ajwfrost Not including 19848840's ane issue (which I don't understand) Looks like the last big issue was resolved 6-7 days ago. Can we expect a release version soon?

**Comment by a9848840** [comment]
@ajwfrost I also hope that the next version can be released directly, mainly to solve the problem of Android 12 garbled characters

**Comment by wmtj** [comment]
We are all waiting for the available version which fixed garbled characters & cut characters issues.

**Comment by a9848840** [comment]
@ajwfrost 

Let me ask you about the current progress, because I really need Android 12 text correction, and please do not crash the air version, I hope the next air version can be released as soon as possible, thank you

**Comment by wmtj** [comment]
@a9848840   31.1.1.795 has been released yestoday .  Link[https://airsdk.harman.com/download/33.1.1.795](url)

**Comment by a9848840** [comment]
thank you!!!
